tools/CopyrightDateTool.scala

```diff
@@ -0,0 +1,171 @@
+    If the message of last change contains "[Sync copyright dates]" it will be ignored and 
+    we will take next commit.
+
+    analyze - just shows if anything is ok
+    update - file fixing
+
+    From root folder:
+    Usage: scala tools/CheckDates.scala analyze
+           scala tools/CheckDates.scala update
+
+  """
+  trait DateComparison {
+    def fileLastUpdateYear: String
+    def gitLastUpdateYear: String
+    def fileCreationYear: String
+    def gitCreationYear: String
+
+    def isUpdateDateWrong(): Boolean = {
+      fileLastUpdateYear.toInt != gitLastUpdateYear.toInt
+    }
+
+    def isUpdateDateWrongStr(): Boolean = {
+      fileLastUpdateYear != gitLastUpdateYear
+    }
+  }
+
+  case class FileDates(pathToFile: String,
+                      fileLastUpdateYear: String,
+                      gitLastUpdateYear: String,
+                      fileCreationYear: String,
+                      gitCreationYear: String) extends DateComparison
+
+  case class GitInfo(message: String, date: String)
+
+  def readFilesList(): Array[String] = {
+    val git_files = Seq("/bin/sh", "-c", 
+      "git grep --name-only --fixed-strings" + 
+      " '|    (c) 20' -- '*.scala'|grep -v CopyrightDateTool").!!
+    val files_array = git_files.split("\n")
+    files_array
+  }
+
+  def readFile(pathToFile: String): String = {
+    val source = scala.io.Source.fromFile(pathToFile)("UTF-8")
+    val lines = try source.mkString finally source.close()
+    lines
+  }
+
+  def getGitCreationYear(pathToFile: String): String = {
+    val buf_dates = (s"git log   --format=%ci $pathToFile" !!)
+    val dates = buf_dates.split("\\n").distinct
+    val year_column = 0
+    val res = dates.last.split("-")(year_column)
+    res
+  }
+
+  def getGitLastUpdateYear(pathToFile: String): String = {
+    val buf_dates = (s"git log -3  --format=%cd $pathToFile" !!)
+    val buf_messages = (s"git log -3  $pathToFile" !!)
+    val first_empty_elem = 1
+    val messages = buf_messages.split("commit").distinct.drop(first_empty_elem)
+    val dates = buf_dates.split("\\n").distinct
+    val gits_info = (messages, dates).zipped.map{ case(a,b) => GitInfo(a,b)}
+    val res = gits_info.filter(!_.message.contains("[Sync copyright dates]"))
+    val year_column = 4
+    val first = 0
+    res(first).date.split(" ")(year_column)
+  }
+
+  def getYear(pathToFile: String, lower_bound: Int, upper_bound: Int): String = {
+    val file = readFile(pathToFile)
+    val third_line = 2
+    val start_index = file.split("\\n")(third_line).indexOf("(c) 20")
+    val year = file.split("\\n")(third_line)
+                   .substring(start_index + lower_bound,
+                              start_index + upper_bound)
+    year
+  }
+
+  def getFileCreationYear(pathToFile: String): String = {
+    val lb = "(c) ".length()
+    val ub = "(c) 20XX".length()
+    val fileCreationYear = getYear(pathToFile, lb, ub)
+    fileCreationYear
+  }
```

Comment:
Instead of reading the file twice, and resplitting the file, you could just use a regex on the output of git grep. (Which gives path, start year and optionally end year.)

Also beware that some files have multiple lines that appear to match, like scaladoc sources that emit copyrights.

Footnote if you were reading the file, you'd `file.getLines.take(10)` to ensure you're only looking at headers.